### PR TITLE
#7382 Fix broken Buffer Double Locking

### LINKED ISSUE #7382: Concurrency Issue in Queue (multiple consumers, one producer) <= Performance + Bug
Ran into this when trying to add buffering to the `FileChannel` IO approach.

If I run 2 consumers against one producer in the way shown in https://github.com/elastic/logstash/pull/7380 one of the consumers eventually locks up.

The reason for this is the following:

Assuming we're starting from an empty queue both consumers will block the last line of the snippet below:

```java
    public Batch readBatch(int limit) throws IOException {
        Page p;

        lock.lock();
        try {
            while ((p = firstUnreadPage()) == null && !isClosed()) {
                try {
                    notEmpty.await();
```

Now this blocking is eventually unlocked by 

```java
 public long write(Queueable element) throws IOException
```

doing this:

```
            // if the queue was empty before write, signal non emptiness
            // a simple signal and not signalAll is necessary here since writing a single element
            // can only really enable a single thread to read a batch
            if (wasEmpty) { notEmpty.signal(); }
```

This is the only place in the `Queue` code that signals this lock. This means:

* If a consumer calls `readBatch()` and neither drains the queue nor fires a new `readBatch()`, the Queue becomes locked since all other consumers continue to wait for the signal (which is never fired by the `write`  since the `wasEmpty` is always false). => bug(?) reproduced in #7380 
* The performance of multiple consumers in general is suboptimal, __as soon as you start to fall behind all consumption happens from a single consumer only__.

### PR DESCRIPTION
**Pull request by original-brownbear**
Fixes #7382, see explaination below

### CODE REVIEW

**Comment by original-brownbear** [comment]
screen shot 2017-06-08 at 19 58 06

fixed => the blocking queue was implemented wrong.

You need to signal the `notEmpty` condition on every append like `ArrayBlockingQueue` does:

```java

    /**
     * Inserts element at current put position, advances, and signals.
     * Call only when holding lock.
     */
    private void enqueue(E x) {
        // assert lock.getHoldCount() == 1;
        // assert items[putIndex] == null;
        final Object[] items = this.items;
        items[putIndex] = x;
        if (++putIndex == items.length)
            putIndex = 0;
        count++;
        notEmpty.signal();
    }
```

More on this can be found under the "bound buffer" problem all over the internet :)

**Comment by colinsurprenant** [comment]
For the record 
> fixed => the blocking queue was implemented wrong.

I would rephrase that for a little bit more precision and contextual information to help anyone reading this to avoid having to ingest the whole #7382 which lead to this fix:

This fixes a concurrency bug in the management of the queue empty state. Specifically the following sequence could lead to a deadlock:

A- reader 1 waits on notEmpty
B- reader 2 waits on notEmpty
C- writer acquires lock, write element and signals notEmpty, lock is unlocked 
D- a reader is selected to be waked up and tries to acquire the lock
E- a writer is "simultaneously" trying to acquire the lock and win in (F)
F- writer acquires lock, write element and **does not** signals notEmpty because queue is not empty at this point
G- readers are blocked.

I agree that the correct fix is to simply remove the condition when signalling non-emptyness 
```
if (wasEmpty) { notEmpty.signal(); }
```
to
```
notEmpty.signal();
```

Note that it seems we have never witnessed this phenomenon in logstash usage probably because at a higher level the reader and writers are also locked. 

I believe we should also backport this into the 5.4 branch for if a minor release is done on 5.4. Even if it seem to not occur in normal logstash usage, I would play safe here. 

LGTM.

**Comment by suyograo** [comment]
yeah, I would backport it to 5.4 branch as well.

**Comment by original-brownbear** [comment]
@suyograo @colinsurprenant 5.4 version pushed https://github.com/elastic/logstash/commit/047291db6954ad8149f9b32cfb8bbdff5171688d :)
